services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

```diff
@@ -0,0 +1,251 @@
+//            ipState.setPortId(portId);
+//            ipState.setSubnetCidr(inSubnetState.getCidr());
+//
+//            String ipManagerServiceUrl = ipUrl + inSubnetState.getId() + "/routes"; // for kubernetes test
+//            HttpEntity<IPStateJson> ipRequest = new HttpEntity<>(new IPStateJson(ipState));
+//            IPStateJson ipResponse = restTemplate.postForObject(ipManagerServiceUrl, ipRequest, IPStateJson.class);
+//            if (ipResponse == null) {
+//                throw new ResourcePersistenceException();
+//            }
+
+            // set up value of properties for subnetState
+            List<RouteWebObject> routes = subnetState.getRoutes();
+            if (routes == null) {
+                routes = new ArrayList<>();
+            }
+            routes.add(routeResponse.getRoute());
+            subnetState.setRoutes(routes);
+            //subnetState.setGatewayIp(ipResponse.getIpState().getIp());
+
+        } catch (ResourceNullException e) {
+            // Route info of subnet rollback
+            List<RouteWebObject> routes = resource.getSubnet().getRoutes();
+            if (routes != null) {
+                RouteWebObject route = routes.get(routes.size() - 1);
+                this.subnetService.routeRollback(route.getId(), resource.getSubnet().getVpcId());
+            }
+            throw new Exception(e);
+        }
+
+        return new SubnetStateJson(subnetState);
+    }
+
+    @RequestMapping(
+            method = PUT,
+            value = {"/project/{projectid}/vpcs/{vpcid}/subnets/{subnetid}", "v4/{projectid}/vpcs/{vpcid}/subnets/{subnetid}"})
+    public SubnetStateJson updateSubnetState(@PathVariable String projectid, @PathVariable String vpcid, @PathVariable String subnetid, @RequestBody SubnetStateJson resource) throws Exception {
+
+        SubnetState subnetState = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(vpcid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(subnetid);
+            SubnetState inSubnetState = resource.getSubnet();
+            RestPreconditionsUtil.verifyResourceNotNull(inSubnetState);
+            RestPreconditionsUtil.populateResourceProjectId(inSubnetState, projectid);
+            RestPreconditionsUtil.populateResourceVpcId(inSubnetState, vpcid);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(subnetid);
+            if (subnetState == null) {
+                throw new ResourceNotFoundException("Subnet not found : " + subnetid);
+            }
+
+            RestPreconditionsUtil.verifyParameterEqual(subnetState.getProjectId(), projectid);
+            RestPreconditionsUtil.verifyParameterEqual(subnetState.getVpcId(), vpcid);
+
+            this.subnetDatabaseService.addSubnet(inSubnetState);
+            subnetState = this.subnetDatabaseService.getBySubnetId(subnetid);
```

Comment:
Do we need to call other microservices here?

Reply:
I don't think so, I will add a new UT to test it.

Reply:
When a customer updates a subnet, if some resources in the subnet (for example, gateway IP) are changing, we might need to call other microservices to make changes accordingly.